Documenting the usage of return type for programmatic navigation (#1480)
vaadin/docs: articles/routing/index.asciidoc

They are typically composed of other smaller components and help you structure and re-use your UI code.
Routing in Vaadin means mapping views into specific URLs and maintaining a meaningful https://en.wikipedia.org/wiki/Deep_linking[deep linking].
A <<route#,route>> in Vaadin lingo means a view mapped to a specific URL. 
URLs can also contain <<route-parameters#,Route Parameters>> or <<additional-guides/query-parameters#,Query Parameters>> that views can utilize to configure their internal state further.
Routing in your application allows your users (by copy-pasting the current URL from their browser) and other applications to share direct links to a specific view and state in your Vaadin UI.

The user <<navigation#,navigates between views>> in your application by interacting with UI elements such as buttons, links, and menus.
Vaadin maintains the proper deep linking URL in your end user's browser when using the framework's built-in mechanisms for navigation.

Routing and navigation should not be considered as the primary mechanism to interact with different parts of your UI, but you should simply define pure Java APIs instead.

> In fact, if deep linking is not needed, even a large Vaadin application can ignore routing and navigation for simplicity and only route the entry view to the root URL of the application.

Reviewer: The message that these two sentences attempt to convey could be clarified more. Do you here mean that user interaction in Vaadin applications is often better achieved by dynamically adding and removing components/views (instead of routing between different views)?

> However, Routing and Navigation APIs in Vaadin help you achieve a decent level of deep linking relatively easily. Thus, we suggest routing at least the top-level views to different URLs and then utilizing more advanced features for views where you need a better deep linking.

Reviewer: > and then utilizing more advanced features for views where you need a better deep linking

It's not clear here what "more advanced features" is referring to.

> In most non-trivial applications, views also need to interact with each other and share data. When initiating navigation, the previous view can <<navigation#passing-data-using-route-parameters,submit trivial data>> as route or query parameters to the new view. The easiest, safest, and performant way to share data between views is to <<navigation#interacting-directly-with-the-target-view,call their Java API directly>>. Doing this, however, will not automatically maintain the URL in the browser. The actual UI state can also lose parity with the URL shown in the browser by internal interaction within the view or, for example, by a message arriving through another thread. In these cases, you should <<route-parameters#maintaining-route-parameters-without-navigation,manually maintain the URL>> if deep linking is required.

Reviewer: ```suggestion
In most non-trivial applications, views also need to interact with each other and share data. When initiating navigation, the previous view can <<navigation#passing-data-using-route-parameters,submit trivial data>> as route or query parameters to the new view. The easiest, safest, and performant way to share data between views is to <<navigation#interacting-directly-with-the-target-view,call their Java API directly>>.
```

The caveat that starts with _"Doing this, however, will not automatically ...."_  is many steps ahead (it assumes the reader has already a good understanding of the Java API for navigation). I think it's better to remove this since the same caveat (with the link to `route-parameters#maintaining-route-parameters-without-navigation`) is already mentioned in the linked Java API for navigation.

= Routing and Navigation

Views in Vaadin are larger components that present logical parts of your UI.

> They are typically composed of other smaller components and help you structure and re-use your UI code.

Reviewer 2: ```suggestion
They are typically composed of other smaller components and help you structure and reuse your UI code.
```

> Routing in Vaadin means mapping views into specific URLs and maintaining a meaningful https://en.wikipedia.org/wiki/Deep_linking[deep linking].

Reviewer 2: ```suggestion
Routing in Vaadin means mapping views into specific URLs and maintaining meaningful https://en.wikipedia.org/wiki/Deep_linking[deep linking].
```

Deep linking allows your users (by copy-pasting the current URL from their browser) and other applications to share direct links to a specific view and state in your Vaadin UI.

A <<route#,route>> in Vaadin lingo means a view mapped to a specific URL. 

> URLs can also contain <<route-parameters#,Route Parameters>> or <<additional-guides/query-parameters#,Query Parameters>> that views can utilize to configure their internal state further.

Reviewer 2: ```suggestion
URLs can also contain <<route-parameters#,route parameters>> or <<additional-guides/query-parameters#,query parameters>> that views can utilize to configure their internal state further.
```

The user <<navigation#,navigates between views>> in your application by interacting with UI elements such as buttons, links, and menus.
Vaadin maintains the proper deep linking URL in your end user's browser when using the framework's built-in mechanisms for navigation.

Routing and navigation is not intended to be the primary mechanism to interact with different parts of your UI or to maintain the state of the application.

> Directly changing the state of components and adding/removing them from the UI using Java API is easier and more efficient.

Reviewer 2: ```suggestion
Directly changing the state of components and adding/removing them from the UI using Java APIs is easier and more efficient.
```

In fact, if deep linking is not needed, even a large Vaadin application can ignore routing and navigation APIs for simplicity and only route the entry view to the root URL of the application.
However, routing and navigation APIs in Vaadin help you achieve a decent level of deep linking relatively easily. 

> Thus, we suggest routing at least the top-level views to different URLs and then utilizing more advanced features like parameters and <<layout#,nested routes>> for views where you need a better deep linking.

Reviewer 2: ```suggestion
Thus, we suggest routing at least the top-level views to different URLs and then utilizing more advanced features like parameters and <<layout#,nested routes>> for views where you need better deep linking.
```

> Sometimes a proper deep linking requires developer to manually <<updating-url-parameters#,maintain the URL parameters>>.

Reviewer 2: ```suggestion
Sometimes proper deep linking requires the developer to manually <<updating-url-parameters#,maintain the URL parameters>>.
```

For example, the UI state can lose parity with the URL shown in the browser by user interaction within the view, like changing search filters, or by a message arriving through another thread.

> Also, other views can be <<navigation#interacting-directly-with-the-target-view,passing data directly>> using Java API, instead of using URL parameters.

Reviewer 2: ```suggestion
Also, other views can be <<navigation#interacting-directly-with-the-target-view,passing data directly>> using Java APIs, instead of using URL parameters.
```

> In these cases the routing mechanism can't update the URL automatically.

Reviewer 2: ```suggestion
In these cases, the routing mechanism cannot update the URL automatically.
```
